File: ios/Runner/Signalling/Signalling.swift
import Foundation
import OpenTok

class Signalling: NSObject {
    
    lazy var publisher: OTPublisher = {
        let settings = OTPublisherSettings()
        settings.name = UIDevice.current.name
        return OTPublisher(delegate: self, settings: settings)!
    }()
    
    var appDelegate: AppDelegate?
    var signal_type = "text-signal"
    
    var session: OTSession?
    
    init(delegate: AppDelegate){
        appDelegate = delegate
    }
    
    func initSession(apiKey: String, sessionId: String, token: String) {
        var error: OTError?
        defer {
            // todo
        }
        
        notifyFlutter(state: SdkState.wait)
        session = OTSession(apiKey: apiKey, sessionId: sessionId, delegate: self)!
        session?.connect(withToken: token, error: &error)
    }
    
    func notifyFlutter(state: SdkState) {
        appDelegate?.vonageChannel?.invokeMethod("updateState", arguments: state.rawValue)
    }
    
    func sendMessage(message: String) {
        session?.signal(withType: signal_type, string: message, connection: nil, error: nil)
    }
    
    func showMessage(message: String, remote: Bool) {
        let arguments = ["message" : message, "remote": remote] as [String : Any]
        appDelegate?.signallingChannel?.invokeMethod("updateMessages", arguments: arguments)
    }
    
}

extension Signalling: OTSessionDelegate {
    
    func sessionDidConnect(_ sessionDelegate: OTSession) {
        print("The client connected to the session.")
        
        var error: OTError?
        self.session?.publish(self.publisher, error: &error)
        
        notifyFlutter(state: SdkState.loggedIn)
    }
    
    func sessionDidDisconnect(_ session: OTSession) {
        print("The client disconnected from the session.")
        notifyFlutter(state: SdkState.loggedOut)
    }
    
    func session(_ session: OTSession, didFailWithError error: OTError) {
        print("The client failed to connect to the session: \(error).")
    }
    
    func session(_ session: OTSession, streamCreated stream: OTStream) {
        print("A stream was created in the session.")
    }
    
    func session(_ session: OTSession, streamDestroyed stream: OTStream) {
        print("A stream was destroyed in the session.")
    }
}

extension Signalling: OTPublisherDelegate {
    func publisher(_ publisher: OTPublisherKit, streamCreated stream: OTStream) {
    }
    
    func publisher(_ publisher: OTPublisherKit, streamDestroyed stream: OTStream) {
    }
    
    func publisher(_ publisher: OTPublisherKit, didFailWithError error: OTError) {
        print("Publisher failed: \(error.localizedDescription)")
    }
    
    func session(_ session: OTSession, receivedSignalType type: String?, from connection: OTConnection?, with string: String?) {
        var isSelf = false
        if (connection?.connectionId == session.connection?.connectionId) {
            isSelf = true
        }
        showMessage(message: string!, remote: isSelf)
        
    }
}

File: android/app/src/main/kotlin/com/vonage/tutorial/opentok/opentok_flutter_samples/singaling/Signalling.kt
package com.vonage.tutorial.opentok.opentok_flutter_samples.singaling

import android.util.Log
import com.opentok.android.OpentokError
import com.opentok.android.Session
import com.opentok.android.Session.SessionListener
import com.opentok.android.Session.SignalListener
import com.opentok.android.Stream
import com.vonage.tutorial.opentok.opentok_flutter_samples.MainActivity


class Signalling(mainActivity: MainActivity) {
    val tag = "Signalling"
    private val signal_type = "text-signal"

    private var activity: MainActivity? = null
    private var session: Session? = null

    private val sessionListener: SessionListener = object : SessionListener {
        override fun onConnected(session: Session) {
            Log.i(tag, "Session Connected")
        }

        override fun onDisconnected(session: Session) {
            Log.i(tag, "Session Disconnected")
        }

        override fun onStreamReceived(session: Session?, stream: Stream?) {
            Log.i(tag, "Stream Received")
        }

        override fun onStreamDropped(session: Session?, stream: Stream?) {
            Log.i(tag, "Stream Dropped")
        }

        override fun onError(session: Session, opentokError: OpentokError) {
            Log.e(tag, "Session error: " + opentokError.message)
        }
    }

    private val signalListener =
        SignalListener { session, type, data, connection ->
            val remote = !connection.equals(session.connection)
            if (type != null && type == signal_type) {
                showMessage(data, remote)
            }
        }

    init {
        activity = mainActivity
    }

    fun initSession(apiKey: String, sessionId: String, token: String) {
        session = Session.Builder(activity, apiKey, sessionId).build()
        session?.setSessionListener(sessionListener)
        session?.setSignalListener(signalListener);
        session?.connect(token)
    }


    fun sendMessage(message: String) {
        Log.d(tag, "Send Message")
        session!!.sendSignal(signal_type, message)
    }

    private fun showMessage(messageData: String, remote: Boolean) {
        Log.d(tag, "Show Message")
        val arguments: HashMap<String, Any> = HashMap()
        arguments["message"] = messageData
        arguments["remote"] = remote
        activity?.updateFlutterMessages(arguments, activity!!.signallingMethodChannel)
    }
}